File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        //staircase searching
        int n = matrix.length;
        int m = matrix[0].length;

        int row = 0;
        int col = m-1;

        while(row < n && col >= 0) {
            int el = matrix[row][col];
            if(el == target) return true;
            else if(el > target) col--;
            else row++;
        }

        return false;
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();

        int sr = 0, er = m-1;
        while(sr <= er) {
            int midR = sr + (er-sr)/2;
            if(matrix[midR][n-1] < target) sr = midR+1;
            else if(matrix[midR][0] > target) er = midR-1;
            else {
                int sc = 0, ec = n-1;
                while(sc <= ec) {
                    int midC = sc + (ec-sc)/2;
                    if(matrix[midR][midC] < target) sc = midC + 1;
                    else if(matrix[midR][midC] > target) ec = midC - 1;
                    else {
                        return true;
                    }
                }
                return false;
            }
        }

        return false;
    }
};